src/Core/Services/PclCryptoFunctionService.cs

```diff
@@ -172,6 +172,55 @@ public Task<byte[]> RsaDecryptAsync(byte[] data, byte[] privateKey, CryptoHashAl
             return Task.FromResult(CryptographicEngine.Decrypt(cryptoKey, data));
         }
 
+        public async Task<byte[]> RsaDecryptAsync(string encValue, byte[] privateKey)
+        {
+            var headerPieces = encValue.Split('.');
+            EncryptionType? encType = null;
+            string[] encPieces = null;
+
+            if (headerPieces.Length == 1)
+            {
+                encType = EncryptionType.Rsa2048_OaepSha256_B64;
+                encPieces = new string[] { headerPieces[0] };
+            }
+            else if (headerPieces.Length == 2 && Enum.TryParse(headerPieces[0], out EncryptionType type))
+            {
+                encType = type;
+                encPieces = headerPieces[1].Split('|');
+            }
+
+            if (!encType.HasValue)
+            {
+                throw new Exception("encType unavailable.");
+            }
+            if (encPieces == null || encPieces.Length == 0)
+            {
+                throw new Exception("encPieces unavailable.");
+            }
+
+            var data = Convert.FromBase64String(encPieces[0]);
+            if (privateKey == null)
+            {
+                throw new Exception("No private key.");
+            }
+
+            var alg = CryptoHashAlgorithm.Sha1;
+            switch (encType.Value)
+            {
+                case EncryptionType.Rsa2048_OaepSha256_B64:
+                case EncryptionType.Rsa2048_OaepSha256_HmacSha256_B64:
+                    alg = CryptoHashAlgorithm.Sha256;
+                    break;
+                case EncryptionType.Rsa2048_OaepSha1_B64:
+                case EncryptionType.Rsa2048_OaepSha1_HmacSha256_B64:
+                    break;
+                default:
+                    throw new Exception("encType unavailable.");
+            }
+
+            return await RsaDecryptAsync(data, privateKey, alg);
+        }
```

Comment:
This is the same as `CryptoService -> RsaDecryptAsync(string encValue)` but with an extra parameter. Please don't duplicate code and instead in the one in `CryptoService`, add the parameter `byte[] privateKey` with `null` as default in the parameter, i.e. `byte[] privateKey = null`; and inside you can have:
```c#
...
var data = Convert.FromBase64String(encPieces[0]);
if(privateKey is null)
{
    privateKey = await GetPrivateKeyAsync();
}
if (privateKey is null)
{
    throw new Exception("No private key.");
}
...
```
Then you can call it with or without the `privateKey` and everything works with this minimum change on the already existing method.
Also if you can, add more verbosity to the exception messages

Reply:
You are totally right i copied the code to test decryption and forgot to refactor it, good catch!